### ISSUE TITLE
C# output overwrites generated output by proto file name

### ISSUE THREAD

**Issue by shaunco** [bufbuild/buf#203]
C# output isn't scoped by package names or directory structure, so if two proto files in different packages have the same file name in the same project, the generated C# output overwrites any previous output:
* `test/first/sample.proto` - with `package test.first` and any set of messages
* `test/second/sample.proto` - with `package test.second` and any set of messages
* `buf.gen.yaml` - set to generate C#

Run `buf generate -o gen` and only a single `gen/csharp/Sample.cs` results. Using a file system monitor, it looks like both first and second were written, but second overwrites first with no warning or error from buf or protoc. `buf check lint` also didn't produce any warnings.

### Sample Project
[buf-csharp-test.zip](https://github.com/bufbuild/buf/files/5595133/buf-csharp-test.zip)

**Comment by bufdev** [comment]
Please do not file duplicate issues with the exact same text but

**Comment by shaunco** [comment]
They are not duplicates. Javascript overwrites by message name and C# by proto filename, which implies totally different places in the internal compiler.

**Comment by shaunco** [comment]
And since you asked in #202, here is the breakdown for #203...

Not sure how this is better than the zip I attached, but here it is:

**buf.yaml**
```yaml
version: v1beta1
build:
  roots:
    - .
lint:
  use:
    - BASIC
    - FILE_LOWER_SNAKE_CASE
  except:
    - ENUM_NO_ALLOW_ALIAS
    - IMPORT_NO_PUBLIC
    - PACKAGE_AFFINITY
    - PACKAGE_DIRECTORY_MATCH
    - PACKAGE_SAME_DIRECTORY
breaking:
  use:
    - WIRE_JSON
```

**buf.gen.yaml**
```yaml
version: v1beta1
plugins:
  - name: csharp
    out: csharp
```

**test/first/sample.proto**
```proto
syntax = "proto3";

package test.first;

option java_multiple_files = true;
option java_package = "org.test.first";
option csharp_namespace = "Test.First";
option go_package = "test.com/first;first";

message Something1 {
    string test_first = 1;
}
```

**test/second/sample.proto**
```proto
syntax = "proto3";

package test.second;

option java_multiple_files = true;
option java_package = "org.test.second";
option csharp_namespace = "Test.Second";
option go_package = "test.com/second;second";

message Something2 {
    string test_second = 1;
}
```

**Comment by bufdev** [comment]
Same as #202, this is due to how the `csharp` builtin plugin works, making compiling per-directory not possible (as it should be with `protoc` plugins). The same thing happens with `protoc`, except it errors:

```
protoc -I . --csharp_out=gen $(find . -name '*.proto')
Sample.cs: Tried to write the same file twice.
```

A mitigation step would be to have different filenames. We'll update buf to correctly display this error (buf not displaying this error was a bug).

**Comment by bufdev** [comment]
> They are not duplicates. Javascript overwrites by message name and C# by proto filename, which implies totally different places in the internal compiler or that message name based logic is in the js generator and the file name based logic is in the C# generator.

This question should be directed to the Protobuf team maintaining the csharp plugin, this is not a buf issue. Buf does not maintain the `js` or `csharp` plugins.

**Comment by shaunco** [comment]
I'll do some digging in to how [Grpc.Tools](https://github.com/grpc/grpc/tree/master/src/csharp/Grpc.Tools) keeps the same directory structure for the C# output as the proto input. I'm guessing they call protoc for each individual file. Seems like it would make a nice Buf option.

**Comment by bufdev** [comment]
This is released in v0.33.0. See https://docs.buf.build/generate-usage for details. Let us know if this works!

**Comment by corlettb** [comment]
Adding "opt: base_namespace" to the buf.gen.yaml for csharp seems to resolve the issue.
